### ISSUE TITLE
How can i adjust the model to my task?

### ISSUE THREAD

**Issue by RuthPikovsky** [ultralytics/ultralytics#18441]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello!
I want to train a keypoint detection model (currently using yolov8). My dataset consists of images with one class 
(one class per image and one class overall, in the future I might want more classes but this is what i
have for now), keypoints and a third parameter that represents a some feature relevant to that keypoint (and I don't really care about the bounding box). It's a float in the scale of 0.0 to 100.0. I want the model to learn the location of the keypoint and the parameter. From my understanding, the "visibility" parameter can do something similar to my regression parameter by replacing the visibility with the value (like the data for the point will look like [0.2, 0.5, 14.0] where the first two are the location of the keypoint and the third is the parameter). Can I make sure that the model also learns my parameter? Do I need a custom loss function for that? 
Another problem I have is that I have a varying number of keypoints per image. it could have 1 keypoint or 350. I understood that the model expects a fixed number of keypoints and to solve that problem I padded the labels with zeros in images where I don't have keypoints (looks like [0.0, 0.0, 0.0]). Is that the correct approach? If so, do I need a custom loss for that aswell so the model doesn't learn the zeros?

Thank you so much,
Ruth

**Comment by UltralyticsAssistant** [comment]
👋 Hello @RuthPikovsky, thank you for your interest in Ultralytics 🚀! We recommend checking out the [Ultralytics Docs](https://docs.ultralytics.com) for guidance on training, where you can explore topics like [model customization](https://docs.ultralytics.com/usage/python/) and using [Python](https://docs.ultralytics.com/usage/python/) or [CLI](https://docs.ultralytics.com/usage/cli/) techniques. Your question raises some interesting points, so let's provide a few pointers:

If this is a 🐛 Bug Report or requires debugging, we suggest sharing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to better understand your dataset structure and any anomalies you're encountering.

For custom training ❓ Questions, like your situation, additional details can help. For example, please include representative dataset images, annotations, and any relevant training logs. Also, verifying that you’re considering our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) is highly recommended.

Some specific notes on your question:  

- Regarding having your regression parameter (the third value), it’s possible to use custom data formats and loss functions. However, it might require some experimentation and understanding of how Ultralytics models handle these inputs.  
- Padding with zeros for variable numbers of keypoints is an interesting approach. Depending on your setup, a custom loss function may indeed be necessary to ensure the model does not learn padded zeros as valid data points.

Feel free to join the vibrant Ultralytics community for more support:  
🎧 [Discord](https://discord.com/invite/ultralytics) - Real-time Q&A and discussions  
📚 [Discourse](https://community.ultralytics.com) - In-depth topic exploration  
💬 [Subreddit](https://reddit.com/r/Ultralytics) - Share insights and experiences  

## Upgrade

Updating to the latest version of the `ultralytics` package might include features or bugfixes relevant to your question. Use the command below:

```bash
pip install -U ultralytics
```

And ensure you meet the [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/).

## Environments

Try out YOLO in these verified environments for seamless training and experimentation:

- **Notebooks** with built-in GPU support: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
- **Google Cloud** and **AWS**: Follow our [GCP Quickstart](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) and [AWS Quickstart](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/) guides.
- **Docker**: Use our [Docker Quickstart](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) to get started easily. <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>  
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) checks are passing, indicating stable operations across tasks and environments.

😊 This is an automated response to provide immediate support. An Ultralytics engineer will follow up soon to address your specific questions in more detail. Stay tuned!

**Comment by Y-T-G** [comment]
> Hello!
> I want to train a keypoint detection model (currently using yolov8). My dataset consists of images with one class 
> (one class per image and one class overall, in the future I might want more classes but this is what i
> have for now), keypoints and a third parameter that represents a some feature relevant to that keypoint (and I don't really care about the bounding box). It's a float in the scale of 0.0 to 100.0. I want the model to learn the location of the keypoint and the parameter. From my understanding, the "visibility" parameter can do something similar to my regression parameter by replacing the visibility with the value (like the data for the point will look like [0.2, 0.5, 14.0] where the first two are the location of the keypoint and the third is the parameter). Can I make sure that the model also learns my parameter? Do I need a custom loss function for that? 

You need to have bounding box labels because the architecture requires it. The visibility value is tied to the visibility as the name says. It probably can't be repurposed to different thing without changing the code.


> Another problem I have is that I have a varying number of keypoints per image. it could have 1 keypoint or 350. I understood that the model expects a fixed number of keypoints and to solve that problem I padded the labels with zeros in images where I don't have keypoints (looks like [0.0, 0.0, 0.0]). Is that the correct approach? If so, do I need a custom loss for that aswell so the model doesn't learn the zeros?

Each keypoint on an image is a class of its own, not some arbitrary point. There has to be consistency between images about what each keypoint represents. Like the keypoint for the left eye should always be assigned to left-eye for every image. Without that consistency and uniqueness, the model wouldn't learn anything and padding wouldn't help.

**Comment by RuthPikovsky** [comment]
Thank you! based on your feedback I came up with a new approach where I will label my data so that each keypoint will be predicted using a small bounding box around it, the center of the box is the location, width and height are small and represent a location where the point can be and the feature is some regression parameter. basically now my data looks like:
0 0.653448 0.000862 0.000862 0.000862 0.242491
0 0.612069 0.033621 0.000862 0.000862 0.171062

where the first value is the class, the next two are x and y, the next two are width and height of the bbox (and they are very small because I'm predicting a small point) and the last is the feature. And we have a bunch of those per image because I have maybe points. The problem becomes different but now I don't have to pad. Is this a better approach?
